<launch>
    <arg name="sim_robot" default="true"/>
    <arg name="robot_ip" default="" unless="$(arg sim_robot)"/>
    <arg name="sim_sensor" default="false"/>
    <arg name="single_camera" default="false" unless="$(arg sim_sensor)"/>

    <arg name="rviz" default="true"/>
    <arg name="box_side" default="0.2"/>
    <arg name="box_x" default="0.15"/>
    <arg name="box_y" default="0.4"/>
    <arg name="box_parent_link" default="workcell_base"/>
    <arg name="plotting" default="false" />
    <arg name="file_write_cb" default="false" />
    <arg name="actuate_gripper" default="true" />
    <arg name="steps_per_phase" default="7" />

    <!-- Bringup the robot -->
    <include file="$(find pick_and_place)/launch/bringup_robot.launch" >
        <arg name="sim_robot" value="$(arg sim_robot)"/>
        <arg name="robot_ip" value="$(arg robot_ip)" unless="$(arg sim_robot)"/>
        <arg name="rviz" value="$(arg rviz)"/>
        <arg name="actuate_gripper" value="$(arg actuate_gripper)"/>
    </include>

    <!-- Publish the location of the camera -->  <!--back, front,right-->
    <node pkg="tf" type="static_transform_publisher" name="camera1_tf_broadcaster"
       args="0.493275 0.309582 1.02424 -1.47137 1.31709 0.0871872 base_link camera1_link 10" />
    <node pkg="tf" type="static_transform_publisher" name="camera2_tf_broadcaster"
       args="0.421917 -0.291741 1.02999 1.62168 1.33258 0.0747153 base_link camera2_link 10" />
    <node pkg="tf" type="static_transform_publisher" name="camera3_tf_broadcaster"
       args="-0.0193177 -0.0140855 1.06183 0.0641083 1.23664 0.0291229 base_link camera3_link 10" />

    <!-- Launch simulated sensor -->
    <group if="$(arg sim_sensor)">
        <!-- Rotate the camera to suit simulation -->
       <node pkg="tf" type="static_transform_publisher" name="camera1_tf_remap"
           args="0 0 0 0 1.57079 0 camera1_link sim_camera1_link 1" />
       <node pkg="tf" type="static_transform_publisher" name="camera2_tf_remap"
           args="0 0 0 0 1.57079 0 camera2_link sim_camera2_link 1" />
       <node pkg="tf" type="static_transform_publisher" name="camera3_tf_remap"
           args="0 0 0 0 1.57079 0 camera3_link sim_camera3_link 1" />

       <!--- Launch 3 camera simulators -->
       <node name="sensor_sim1" pkg="pick_and_place" type="sensor_simulator_3d" args="/camera1/camera/depth_registered/points sim_camera1_link" output="screen"/>
       <node name="sensor_sim2" pkg="pick_and_place" type="sensor_simulator_3d" args="/camera2/camera/depth_registered/points sim_camera2_link"/>
       <node name="sensor_sim3" pkg="pick_and_place" type="sensor_simulator_3d" args="/camera3/camera/depth_registered/points sim_camera3_link"/>
    </group>

    <!-- Set simulated box parameters -->
    <param name="box_side" value="$(arg box_side)"/>
    <param name="box_x" value="$(arg box_x)"/>
    <param name="box_y" value="$(arg box_y)"/>
    <param name="box_parent_link" value="$(arg box_parent_link)"/>

    <!-- Launch real sensor -->
    <group unless="$(arg sim_sensor)">
         <!-- Launch camera node -->
         <include file="$(find pick_and_place)/launch/bringup_realsense.launch" >
            <arg name="single_camera" value="$(arg single_camera)" />
         </include>
    </group>

    <!-- Launch vision node -->
    <include file="$(find pick_and_place_perception)/launch/processing_node.launch" />

    <!-- Launch planner node -->
    <node name="pick_and_place_node" pkg="pick_and_place" type="pick_and_place_node" output="screen">
            <param name="sim_robot" value="$(arg sim_robot)"/>
            <param name="actuate_gripper" value="$(arg actuate_gripper)"/>
            <param name="plotting" value="$(arg plotting)"/>
            <param name="file_write_cb" value="$(arg file_write_cb)" />
            <param name="steps_per_phase" value="$(arg steps_per_phase)" />
    </node>

    <!-- Launch visualization -->
    <node if="$(arg rviz)" pkg="rviz" type="rviz" name="pick_and_place_viz"
        args="-d $(find pick_and_place)/config/pick_and_place.rviz" />

</launch>

